app/src/parser_utils.c

```diff
@@ -0,0 +1,404 @@
+#include "parser_utils.h"
+
+#include <zxmacros.h>
+
+#include "parser_special.h"
+#include "parser_txdef.h"
+#include "runtime_arg.h"
+
+parser_error_t readU64(parser_context_t *ctx, uint64_t *result) { return _readUInt64(ctx, result); }
+
+parser_error_t readU32(parser_context_t *ctx, uint32_t *result) { return _readUInt32(ctx, result); }
+
+parser_error_t readU16(parser_context_t *ctx, uint16_t *result) { return _readUInt16(ctx, result); }
+
+parser_error_t readU8(parser_context_t *ctx, uint8_t *result) { return _readUInt8(ctx, result); }
+
+parser_error_t parser_init_context(parser_context_t *ctx, const uint8_t *buffer, uint16_t bufferSize) {
+    ctx->offset = 0;
+    ctx->buffer = NULL;
+    ctx->bufferLen = 0;
+
+    if (bufferSize == 0 || buffer == NULL) {
+        // Not available, use defaults
+        return parser_init_context_empty;
+    }
+
+    ctx->buffer = buffer;
+    ctx->bufferLen = bufferSize;
+
+    entry_point_offset = 0;
+
+    return parser_ok;
+}
+
+parser_error_t parser_init(parser_context_t *ctx, const uint8_t *buffer, uint16_t bufferSize) {
+    CHECK_PARSER_ERR(parser_init_context(ctx, buffer, bufferSize))
+    return parser_ok;
+}
+
+bool is_container_type(uint8_t cl_type) {
+    return cl_type == TAG_OPTION || cl_type == TAG_LIST || (cl_type >= TAG_RESULT && cl_type <= TAG_ANY);
+}
+
+bool is_map_type(uint8_t cl_type) { return cl_type == TAG_MAP; }
+
+parser_error_t check_runtime_type(uint8_t cl_type) {
+    if (cl_type > NUM_RUNTIME_TYPES) {
+        return parser_context_unknown_prefix;
+    } else {
+        return parser_ok;
+    }
+}
+
+parser_error_t parse_item(parser_context_t *ctx) {
+    uint32_t part = 0;
+    CHECK_PARSER_ERR(_readUInt32(ctx, &part));
+    if (part >= ctx->bufferLen - ctx->offset) {
+        return parser_unexpected_buffer_end;
+    }
+    ctx->offset += part;
```

Comment:
check buffer len here: 
```suggestion
   if (ctx->offset + part > ctx->bufferLen) {
        return parser_unexpected_buffer_end;
    }
```

Reply:
we should create a fuzzer to validate the new parser

Reply:
update APDUSPEC.md wih the new transactions